### PR TITLE
Add `on_exception` to DataModule

### LINKED ISSUE #19594: Notify LightningDataModule on exception
### Description & Motivation

Callbacks have both `teardown` and `on_exception` to gracefully terminate, but `LightningDataModule` does not have `on_exception` and `teardown` is only seem to be called when fit/validate/predict/test finished successfully.

### Pitch

Can exception cases be also added to `LightningDataModule` contract? Or maybe add `on_exception`. 

### Alternatives

Otherwise the only contract we have is the same one PyTorch `DataLoader` is using with `__del__` and daemon processes: https://github.com/pytorch/pytorch/blob/main/torch/utils/data/dataloader.py#L763

### Additional context

Note that `atexit` won't help if `LightningDataModule` is spawning non-daemon processes I believe.

cc @borda @carmocca @awaelchli @justusschock

### PR DESCRIPTION
**Pull request by clumsy**
## What does this PR do?

Adds `on_exception` callback that lets us handle training termination due to an exception for `LightningDataModule`, which is currently not available.

Fixes #19594

The change is backward-compatible, but could be breaking if a type derived from `LightningDataModule` defines `on_exception` method.

<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--19601.org.readthedocs.build/en/19601/

### CODE REVIEW

**Comment by clumsy** [comment]
Please let me know if additional changes are necessary, @awaelchli

File: src/lightning/pytorch/CHANGELOG.md [hunk 1]
```diff
@@ -14,6 +14,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - Added robust timer duration parsing with an informative error message when parsing fails ([#19513](https://github.com/Lightning-AI/pytorch-lightning/pull/19513))
 
+- Added `on_exception` hook to `LightningDataModule` to gracefully handle the exception that interrupted the training
```
Comment (awaelchli) [inline src/lightning/pytorch/CHANGELOG.md:17]: I would drop the explanation here. The hooks are not actually meant to handle the exception itself, but to run additional code before raising the exception. 
```suggestion
- Added `on_exception` hook to `LightningDataModule`
```

Comment (awaelchli) [reply src/lightning/pytorch/CHANGELOG.md:17]: Also let's add the link to your PR:

```suggestion
- Added `on_exception` hook to `LightningDataModule` ([#19601](https://github.com/Lightning-AI/pytorch-lightning/pull/19601))
```

**Review by awaelchli — APPROVED** [review]
Hey @clumsy 

The implementation looks good! Thanks for adding the test.

**Review by Borda — APPROVED** [review]

**Review by carmocca — APPROVED** [review]
